engines/Collection.py

```diff
@@ -77,4 +72,44 @@ def search(self, **search_args):
         search_response = self.native_search(request=request)
         if "log" in search_args:
             print(json.dumps(search_response, indent=2))
-        return self.transform_response(search_response)
\ No newline at end of file
+        return self.transform_response(search_response)
+
+    def hybrid_search(self, searches=[], limit=None, algorithm="rrf", algorithm_params={}):
+        hybrid_search_results = None
+        match algorithm:
+            case "rrf":
+                search_results = [self.search(**request)["docs"]
+                                  for request in searches]
+
+                hybrid_search_scores = reciprocal_rank_fusion(search_results,
+                                                              algorithm_params.get("k"))
+                scored_docs = self.merge_search_results(search_results, 
+                                                hybrid_search_scores)    
+                return {"docs": scored_docs[:limit]}
+            case "lexical_vector_rerank":
+                lexical_search_request = searches[0]
+                searches[1]["k"] = algorithm_params.get("k", 10) #TODO: should probably default to "limit" instead of 10
+                lexical_search_request["rerank_query"] = searches[1]
+                return self.search(**lexical_search_request)
+        return hybrid_search_results
+
+    def merge_search_results(self, search_results, scores):   
+        merged_results = {}
+        for results in search_results:
+            for doc in results:
+                if doc["id"] in merged_results:
+                    merged_results[doc["id"]] = {**doc, **merged_results[doc["id"]]}
+                else: 
+                    merged_results[doc["id"]] = doc
+        return [{**merged_results[id], "score": score}
+                for id, score in scores.items()]
+    
+    
+def reciprocal_rank_fusion(search_results, k = None):
```

Comment:
Note: I put this at top level (outside of class) since it can be a static method and that way I can load it into the jupyter notebook instead of duplicating. Open to putting it in a different file if you prefer and then importing into the Collection.

Per our discussion yesterday, I also went ahead and moved all the hybrid search methods to the Collection as opposed to the SolrCollection so they can be reused. Any engine that implements these themselves can always override the method, but that's better than forcing each engine that doesn't support to copy/pasts the shared implementation.